File: pcs/jeremy-jeffrey-matos-cangalaya-graded-practise-4/exercise04/solution.py
from cv2 import (
    AKAZE,
    ORB,
    SIFT,
    SURF,
    Affine,
    AffineBestOf2Nearest,
    BestOf2NearestRange,
    CompressedRectilinear,
    CompressedRectilinearPortrait,
    Cylindrical,
    FeatherBlender,
    Fisheye,
    Mercator,
    MultiBandBlender,
    Panini,
    PaniniPortrait,
    Plane,
    Spherical,
    Stereographic,
    TransverseMercator,
)

def stitch_images(
    full_path_input_image,
    blender,  # ={FeatherBlender | MultiBandBlender}
    features_finder,  # ={AKAZE | ORB | SIFT | SURF}
    features_matcher,  # ={AffineBestOf2Nearest | BestOf2NearestRange}
    warper,  # ={ Affine | CompressedRectilinearPortrait | CompressedRectilinear | Cylindrical | Fisheye | Mercator | PaniniPortrait | Panini | Plane | Spherical | Stereographic | TransverseMercator }
    full_path_output_image,
):
    pass


if __name__ == "__main__":
    stitch_images(
        full_path_input_image=[
            "panorama1-input-1.jpg",
            "panorama1-input-2.jpg",
            "panorama1-input-3.jpg",
            "panorama1-input-4.jpg",
            "panorama1-input-5.jpg",
            "panorama1-input-6.jpg",
        ],
        blender=MultiBandBlender,
        features_finder=SIFT,
        features_matcher=BestOf2NearestRange,
        warper=Mercator,
        full_path_output_image="panorama1-mercator.jpg",
    )

"""
    stitch_images(
        full_path_input_image="image-1.jpg",
        blender=FeatherBlender,
        features_finder=AKAZE,
        features_matcher=AffineBestOf2Nearest,
        warper=Affine,
        full_path_output_image="panorama-1.jpg",
    )
    stitch_images(
        full_path_input_image="image-2.jpg",
        blender=MultiBandBlender,
        features_finder=ORB,
        features_matcher=BestOf2NearestRange,
        warper=CompressedRectilinearPortrait,
        full_path_output_image="panorama-2.jpg",
    )
    stitch_images(
        full_path_input_image="image-3.jpg",
        blender=FeatherBlender,
        features_finder=SIFT,
        features_matcher=AffineBestOf2Nearest,
        warper=CompressedRectilinear,
        full_path_output_image="panorama-3.jpg",
    )
    stitch_images(
        full_path_input_image="image-4.jpg",
        blender=MultiBandBlender,
        features_finder=SURF,
        features_matcher=AffineBestOf2Nearest,
        warper=Cylindrical,
        full_path_output_image="panorama-4.jpg",
    )
    stitch_images(
        full_path_input_image="image-5.jpg",
        blender=FeatherBlender,
        features_finder=AKAZE,
        features_matcher=BestOf2NearestRange,
        warper=Fisheye,
        full_path_output_image="panorama-5.jpg",
    )
    stitch_images(
        full_path_input_image="image-6.jpg",
        blender=MultiBandBlender,
        features_finder=ORB,
        features_matcher=AffineBestOf2Nearest,
        warper=Mercator,
        full_path_output_image="panorama-6.jpg",
    )
    stitch_images(
        full_path_input_image="image-7.jpg",
        blender=FeatherBlender,
        features_finder=SIFT,
        features_matcher=BestOf2NearestRange,
        warper=PaniniPortrait,
        full_path_output_image="panorama-7.jpg",
    )
    stitch_images(
        full_path_input_image="image-8.jpg",
        blender=MultiBandBlender,
        features_finder=SURF,
        features_matcher=AffineBestOf2Nearest,
        warper=Panini,
    )
"""

File: pcs/jeremy-jeffrey-matos-cangalaya-graded-practise-4/exercise04/solution.cpp
#include <iostream>
#include <vector>
#include <opencv2/opencv.hpp>
#include <opencv2/stitching.hpp>
#include <opencv2/features2d.hpp>
#include <filesystem>

namespace fs = std::filesystem;

const int FeatherBlender = 0;
const int MultiBandBlender = 1;

const int AKAZE = 0;
const int ORB = 1;
const int SIFT = 2;
const int SURF = 3;

const int AffineBestOf2Nearest = 0;
const int BestOf2NearestRange = 1;

const int Affine = 0;
const int CompressedRectilinearPortrait = 1;
const int CompressedRectilinear = 2;
const int Cylindrical = 3;
const int Fisheye = 4;
const int Mercator = 5;
const int PaniniPortrait = 6;
const int Panini = 7;
const int Plane = 8;
const int Spherical = 9;
const int Stereographic = 10;
const int TransverseMercator = 11;

void stitch_images(
    std::vector<std::string> full_path_input_image,
    int blender,
    int features_finder,
    int features_matcher,
    int warper,
    std::string full_path_output_image)
{
    std::vector<cv::Mat> images;

    for (const std::string &path : full_path_input_image)
    {
        cv::Mat image = cv::imread(path);
        if (image.empty())
        {
            std::cerr << "Could not open or find the image at " << path << std::endl;
            return;
        }
        images.push_back(image);
    }

    cv::Ptr<cv::Stitcher> stitcher = cv::Stitcher::create(cv::Stitcher::PANORAMA);

    cv::Ptr<cv::detail::Blender> b;
    cv::Ptr<cv::Feature2D> ff;
    cv::Ptr<cv::detail::FeaturesMatcher> fm;
    cv::Ptr<cv::WarperCreator> w;

    if (blender == FeatherBlender)
    {
        b = cv::detail::FeatherBlender::createDefault(1);
    }
    else if (blender == MultiBandBlender)
    {
        b = cv::detail::MultiBandBlender::createDefault(2);
    }
    else
    {
        std::cerr << "Unrecognized blender type!" << std::endl;
        return;
    }

    if (features_finder == AKAZE)
    {
        ff = cv::AKAZE::create();
    }
    else if (features_finder == ORB)
    {
        ff = cv::ORB::create();
    }
    else if (features_finder == SIFT)
    {
        ff = cv::SIFT::create();
    }
    else if (features_finder == SURF)
    {
        std::cerr << "not suportted" << std::endl;
        return;
    }
    else
    {
        std::cerr << "Unrecognized features finder type" << std::endl;
        return;
    }

    if (features_matcher == AffineBestOf2Nearest)
    {
        fm = cv::makePtr<cv::detail::AffineBestOf2NearestMatcher>();
    }
    else if (features_matcher == BestOf2NearestRange)
    {
        fm = cv::makePtr<cv::detail::BestOf2NearestRangeMatcher>();
    }
    else
    {
        std::cerr << "Unrecognized features matcher type" << std::endl;
        return;
    }

    float warper_scale = 1.0;

    switch (warper)
    {
    case Affine:
        w = cv::makePtr<cv::AffineWarper>();
        break;
    case CompressedRectilinearPortrait:
        w = cv::makePtr<cv::CompressedRectilinearPortraitWarper>(warper_scale);
        break;
    case CompressedRectilinear:
        w = cv::makePtr<cv::CompressedRectilinearWarper>(warper_scale);
        break;
    case Cylindrical:
        w = cv::makePtr<cv::CylindricalWarper>();
        break;
    case Fisheye:
        w = cv::makePtr<cv::FisheyeWarper>();
        break;
    case Mercator:
        w = cv::makePtr<cv::MercatorWarper>();
        break;
    case PaniniPortrait:
        w = cv::makePtr<cv::PaniniPortraitWarper>(warper_scale);
        break;
    case Panini:
        w = cv::makePtr<cv::PaniniWarper>(warper_scale);
        break;
    case Plane:
        w = cv::makePtr<cv::PlaneWarper>();
        break;
    case Spherical:
        w = cv::makePtr<cv::SphericalWarper>();
        break;
    case Stereographic:
        w = cv::makePtr<cv::StereographicWarper>();
        break;
    case TransverseMercator:
        w = cv::makePtr<cv::TransverseMercatorWarper>();
        break;
    default:
        std::cerr << "Unrecognized warper type!" << std::endl;
        return;
    }

    stitcher->setBlender(b);
    stitcher->setFeaturesFinder(ff);
    stitcher->setFeaturesMatcher(fm);
    stitcher->setWarper(w);

    cv::Mat pano;

    cv::Stitcher::Status status = stitcher->stitch(images, pano);

    if (status != cv::Stitcher::OK)
    {
        std::cerr << "Can't stitch images, error code = " << int(status) << std::endl;
        return;
    }

    // Save the result
    cv::imwrite(full_path_output_image, pano);
}

int main()
{
    std::string path = "./images-for-stitcher/";
    std::vector<std::string> full_path_input_image = {};

    try
    {
        for (const auto &entry : fs::directory_iterator(path))
        {
            if (entry.is_regular_file())
            {
                full_path_input_image.push_back(entry.path().string());
            }
        }
    }
    catch (const fs::filesystem_error &e)
    {
        std::cerr << "Filesystem error: " << e.what() << std::endl;
    }
    catch (const std::exception &e)
    {
        std::cerr << "Exception: " << e.what() << std::endl;
    }

    int blender = MultiBandBlender;
    int features_finder = SIFT;
    int features_matcher = BestOf2NearestRange;
    int warper = Mercator;
    std::string full_path_output_image = "./output.jpg";

    stitch_images(full_path_input_image, blender, features_finder, features_matcher, warper, full_path_output_image);

    return 0;
}